Add ArraySegment Enumerator Tests

AlexRadch opened the pull request:
Add tests for ArraySegment Enumerator. See https://github.com/dotnet/corefx/issues/14170

AlexRadch commented:
@shmao, @zhenlan I have strange serialization error with serialization tests can you help me?

`/mnt/resource/j/workspace/dotnet_corefx/master/centos7.1_debug_prtest/src/Common/tests/System/Collections/IEnumerable.Generic.Serialization.Tests.cs(9,36): error CS0234: The type or namespace name 'Formatters' does not exist in the namespace 'System.Runtime.Serialization' (are you missing an assembly reference?) [/mnt/resource/j/workspace/dotnet_corefx/master/centos7.1_debug_prtest/src/System.Runtime/tests/System.Runtime.Tests.csproj]`

What should I add ?

shmao commented:
It's most likely that your the test project need to reference `System.Runtime.Serialization.Formatters` in project.json.

joperezr commented:
The reason this is failing is because you modified System.Runtime.Tests.csproj to include file `IEnumerable.Generic.Serialization.Tests.cs` which depends on System.Runtime.Serialization.Formatters.Binary and that is only available for `netstandard1.7`. We are x-compiling this test project for `netstandard1.5` and that is why you are seeing this error. The solution would be to condition the addition of that test file so that it only gets includded when building for `netstandard1.7`. To fix that, simply move that line to the section where we include things for the default targetGroup like:
```xml
  <ItemGroup Condition="'$(TargetGroup)' == ''">
    (...)
    <Compile Include="..\..\Common\tests\System\Collections\IEnumerable.Generic.Serialization.Tests.cs">
      <Link>Common\System\Collections\Generic\IEnumerable.Generic.Serialization.Tests.cs</Link>
    </Compile>
    (...)
  </ItemGroup>
```

AlexRadch commented:
@dnlharvey, @AtsushiKan What is wrong with `System.Reflection.Tests.ModuleTests.ResolveTypeFail` test?
I did not change it
```
            System.Reflection.Tests.ModuleTests.ResolveTypeFail(token: 33555511) [FAIL]
       The target "BuildAllProjects" listed in a BeforeTargets attribute at "C:\mount\corefx\Tools\versioning.targets (218,11)" does not exist in the project, and will be ignored.
       RunTestsForProject:
               Assert.Throws() Failure
               Expected: typeof(System.ArgumentException)
               Actual:   (No exception was thrown)
               Stack Trace:
```

AlexRadch commented:
@karelz can you tell who can help me with strange `System.Reflection.Tests.ModuleTests.ResolveTypeFail' failed test? I updated ArraySegment tests and all them is passed. Why ResolveTypeFail failed?

AlexGhiondea commented:
@AlexRadch it looks like the test which expected to fail for a bad token is now able to resolve.

@alexperovich does this sound familiar?

karelz commented:
@AlexRadch did you get the test failure under debugger? Does it repro consistently or is it intermittent failure?

alexperovich commented:
Yea, I added those tests. It tries to use Module.ResolveType on "bad" metadata token values. One is a method metadata token which will always be bad. Something made the value 33555511 resolvable as a type token.

AlexRadch commented:
@alexperovich How I can change this test to pass it?

alexperovich commented:
@AlexRadch change the test input that is failing. Here: https://github.com/dotnet/corefx/blob/master/src/System.Runtime/tests/System/Reflection/ModuleTests.cs#L186.
My guess is that the third entry is the failing one. The 1000 might need to be increased.

AlexGhiondea commented:
@alexperovich if we want a bad token number, why not pick a random number instead?

AlexRadch commented:
@AlexGhiondea I think that random number sometimes will be correct.

AlexRadch commented:
@AlexGhiondea @joperezr I fixed BadResolveTypes issue. Code review please.

joperezr commented:
@dotnet-bot test Innerloop OSX Debug Build and Test please

karelz commented:
@alexperovich do you understand why @AlexRadch's change triggered the test to fail consistently?
If it is unrelated, the test fix should not be part of this PR. If it is related, we need to fully understand why.
Marking as blocked ...

AlexRadch commented:
@karelz I changed `System.Reflection.Tests.ModuleTests.ResolveTypeFail` test to fix issue with it.

@alexperovich answer
> @AlexRadch change the test input that is failing. Here: https://github.com/dotnet/corefx/blob/master/src/System.Runtime/tests/System/Reflection/ModuleTests.cs#L186.
My guess is that the third entry is the failing one. The 1000 might need to be increased.

So I only changed 1000 to 10000 in `ResolveTypeFail` test to pass it as @alexperovich answered. So I think PR should be unblocked.

Also I want to write many other tests to my PRs and all of them will use changes from this PR so I am waiting while this PR will be merged.

What I should change to merge this PR without blocking?

Before it was not merged because it not pass `System.Reflection.Tests.ModuleTests.ResolveTypeFail` test that have issue. After I changed issue in `System.Reflection.Tests.ModuleTests.ResolveTypeFail` test. This PR is blocked.

So now I do not understand how change this PR to merge it.

karelz commented:
> @AlexRadch I changed System.Reflection.Tests.ModuleTests.ResolveTypeFail test to fix issue with it.

I know you did. It, However, it does not explain why it failed in the first place with your changes. That's the key thing missing - we are blocking on that understanding.

AlexRadch commented:
@karelz I see @alexperovich answer that 1000 is too small and should be increased. That I did.

>@AlexRadch change the test input that is failing. Here: https://github.com/dotnet/corefx/blob/master/src/System.Runtime/tests/System/Reflection/ModuleTests.cs#L186.
My guess is that the third entry is the failing one. The 1000 might need to be increased.

I think it because I added serialization tests to `System.Runtime.Tests.csproj` see https://github.com/dotnet/corefx/pull/14431/files#diff-6223985b207d5111dc7d0c2a37cff4e7R142

I do not know why but  `System.Runtime.Tests` did not test any collection on serialization. So 1000 was enough for reflection. Now when I included serialization tests, it is not enough. 

To unblock this PR,  should I remove serialization tests from `System.Runtime.Tests.csproj` to pass old `ResolveTypeFail` test with 1000?

karelz commented:
Ah, I missed the fact that we added more tests into the module - that could have made the +1000 a valid token. That would explain things sufficiently. Removing the blocking status for now ...

@alexperovich can you please confirm?

alexperovich commented:
Yes, that is correct. When we add more types into the module that could make +1000 a valid token.

AlexGhiondea commented:
@alexperovich surely there is a way to fix the test so that we never have to worry about not getting an invalid token when new types are added! 

Can you work with @AlexRadch to figure out such a way?

karelz commented:
Agreed. We should file a separate issue to improve the test. IMO we should not block this PR on it.

AlexGhiondea commented:
Yup -- let's not block on this. 

Filed https://github.com/dotnet/corefx/issues/14625 to track it.

AlexRadch commented:
@AlexGhiondea @joperezr joperezr PR is unblocked. Code review please.

AlexGhiondea commented on src/System.Runtime/tests/System/ArraySegmentTests.cs:
```diff
@@ -5,64 +5,115 @@
+            return new ArraySegment<T>();
+        }
+
+        protected virtual ArraySegment<T> Factory(int count, int offset, int lenght)
```
nit: typo: `lenght` -> `length`

karelz commented:
Thanks @AlexRadch for your contribution!

AlexRadch commented:
Thanks to all!

jamesqo commented:
@karelz Do you know what this PR actually did? I took 2 quick glances but I'm not sure what it's purpose was. It didn't expose the `Enumerator ArraySegment<T>.GetEnumerator()` API for https://github.com/dotnet/corefx/issues/14170, because it's not there in the ref file.

karelz commented:
@jamesqo I think it added just bunch of tests.
